migrations/1686291754032-copy-legacy-tools-to-ctl.js

```diff
@@ -0,0 +1,374 @@
+	return {
+		name: ltiToolTemplate.name,
+		url: ltiToolTemplate.url,
+		logoUrl: ltiToolTemplate.logo_url,
+		parameters: mapToExternalToolParameter(ltiToolTemplate),
+		isHidden: ltiToolTemplate.isHidden,
+		openNewTab: ltiToolTemplate.openNewTab,
+		version: 1,
+		...toolConfigMapper(ltiToolTemplate),
+	};
+}
+
+function mapToSchoolExternalTool(externalTool, ltiToolTemplate, context) {
+	return {
+		tool: externalTool._id,
+		school: context.schoolId,
+		schoolParameters: mapToCustomParameterEntry(externalTool.parameters, ltiToolTemplate.customes),
+		toolVersion: externalTool.version,
+	};
+}
+
+function mapToContextExternalTool(schoolExternalTool, externalTool, context) {
+	return {
+		schoolTool: schoolExternalTool._id,
+		contextId: context.id,
+		contextType: 'course',
+		contextToolName: externalTool.name,
+		parameters: schoolExternalTool.parameters,
+		toolVersion: schoolExternalTool.version,
+	};
+}
+
+// How to use more than one schema per collection on mongodb
+// https://stackoverflow.com/questions/14453864/use-more-than-one-schema-per-collection-on-mongodb
+module.exports = {
+	up: async function up() {
+		await connect();
+
+		// eslint-disable-next-line no-process-env
+		if (process.env.SC_THEME !== 'n21') {
+			info('Migration does not migrate lti tools to CTL for this instance.');
+			return;
+		}
+
+		// FIND ALL LTI TOOL TEMPLATES
+		const ltiToolTemplates = await LtiTool.find({
+			$or: [{ name: { $regex: `Bettermarks` } }, { name: { $regex: `Nextcloud` } }],
+			isTemplate: true,
+		})
+			.lean()
+			.exec();
+
+		if ((ltiToolTemplates || []).length === 0) {
+			alert('No LtiTool Template found.');
+			return;
+		}
+
+		// FIND ALL LEGACY TOOLS
+		const ltiTools = await LtiTool.find({
+			$or: [{ name: { $regex: `Bettermarks` } }, { name: { $regex: `Nextcloud` } }],
+			isTemplate: false,
+		})
+			.lean()
+			.exec();
+
+		/* eslint-disable no-await-in-loop */
+		for (const ltiTool of ltiTools) {
+			// GET TOOLTEMPLATE
+			const toolTemplate = ltiToolTemplates.filter((template) => template.name === ltiTool.name);
+
+			// GET COURSE
+			const course = await Course.findOne({
+				ltiToolIds: { $in: [ltiTool.id] },
+			})
+				.lean()
+				.exec();
+
+			if (course === undefined) {
+				alert(`No course found with LtiToolId: ${ltiTool.id}.`);
+				// eslint-disable-next-line no-continue
+				continue;
+			}
+
+			// GET EXTERNALTOOL
+			let externalTool = await ExternalTool.findOne({
+				name: toolTemplate.name,
+				url: toolTemplate.url, // or other paramter for the query
+			})
+				.lean()
+				.exec();
+
+			// CHECK IF EXTERNALTOOL EXISTS
+			if (externalTool === undefined) {
+				externalTool = mapToExternalTool(toolTemplate);
+				externalTool = await ExternalTool.save(externalTool);
+			}
+
+			let schoolExternalTool;
+
+			// GET SCHOOLEXTERNALTOOL
+			const schoolExternalTools = await SchoolExternalTool.find({
+				school: course.schoolId,
+				tool: externalTool._id,
```

Comment:
school external tools have no name. they are unique regarding their linked external tool id and school id.